messaging/spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -0,0 +1,187 @@
+
+In a microservice based architecture, temporal coupling results in a fragile system, with resilience and scaling properties that are worse than a monolith, hence, it is important for microservice based architectures to embrace asynchronous communication as much as possible. MicroProfile currently does not offer support for asynchronous messaging, only synchronous messaging using JAX-RS and MicroProfile REST client are supported.
+
+Java EE offers JMS and Message Driven Beans for handling asynchronous communication, however, there are some problems with these specifications:
+
+* Both are designed for a technology landscape where messaging was a heavy weight process used on the edge of system to hand control of a transaction from system to another, consequently there is a lot of boilerplate associated with using these technologies.
+* It is assumed their design that consistency is handled using distributed transactions. However, modern message brokers, popular in microservice deployments, such as Kafka, Amazon Kinesis and Azure Event Hubs, do not support XA transactions, rather, message acknowledgement is handled using offsets with at least once delivery guarantees.
+* They do not have support for asynchronous IO, it is assumed that message processing is done on a single thread, however many modern specs are moving to asynchronous IO.
+
+Hence a lighter weight, reactive solution to messaging is desirable for MicroProfile to ensure microservices written using MicroProfile are able to meet the demands required by the architecture.
+
+=== Design
+
+Applications declare message subscribers, publishers and processors by annotating methods on CDI beans with `@Incoming` and `@Outgoing` annotations. A method with an `@Incoming` annotation is a message subscriber, it consumes messages from a messaging provider. A method with an `@Outgoing` annotation is a message publisher, it emits messages to a messaging provider. A method with both an `@Incoming` and an `@Outgoing` annotation is a message processor, it consumes messages from one messaging provider, does some transformation to them, and emits messages to another, potentially different, messaging provider.
+
+==== Messaging providers
+
+A messaging provider is an abstract concept provided by this spec, it typically will correspond to a messaging transport, such as Kafka, Amazon Kinesis, RabbitMQ or ActiveMQ. It is up an implementation as to which and how many transports it will support. Implementations may provide a default messaging transport, and that default may be configurable, all this is left up to the implementation to decide on the mechanism for providing that. Implementations may also provide an SPI for plugging in custom messaging providers, the definition of that SPI is up to the implementation, and beyond the scope of this specification. A future iteration of this spec may provide such an SPI, for example based on the Java EE Connector Architecture, but currently this is out of scope.
```

Comment:
> to a messaging transport, such as Kafka, Amazon Kinesis, RabbitMQ or ActiveMQ 

should these be called as brokers, instead of transport? 

For me transport is more the _messaging transport protocol_, like AMQP, MQTT or the Apache Kafka protocol

Reply:
Good point to raise. I've tried to use transport in the same way that CloudEvents is using it https://github.com/cloudevents/spec/blob/master/spec.md. It does consider transports and protocols to be different things. I consider transport to be a complete system for getting messages from one place to another. Just as flight is a system of transport that includes planes, airports, airlines that schedule planes to fly between particular places at particular times, etc, Kafka is a system of transport that includes a message broker, a protocol for speaking to it, client libraries for communicating with it, etc.

Reply:
Ok, fine w/ that